### ISSUE TITLE
Load issues in DGraph

### ISSUE THREAD

**Issue by ajayktkp** [dgraph-io/dgraph#1914]
- What version of Dgraph are you using?
0.9.4

- Have you tried reproducing the issue with latest release?
already using the latest release

- What is the hardware spec (RAM, OS)?
16GB Ram, 200 GB (SSD), Ubuntu 14.04.2 LTS (Amazon gp2 instance)

- Steps to reproduce the issue (command/config used to run Dgraph).
1. Start with a fresh DGraph installation. 
Commands used to start the DGraph:
/usr/local/bin/dgraph zero --bindall=true -o 1808 --my 10.0.11.162:8888
/usr/local/bin/dgraph server --memory_mb=8188 --bindall=True --my=10.0.11.162:5080 --idx 1 -o -2000 --zero=10.0.11.162:8888 --trace 1.0

2. create schema: 
curl localhost:6080/alter -XPOST -d $'
  name: string @index(term) .
' | python -m json.tool | less

3. Run following go program to insert data:
```
func TestingLoad(outerloop, parallelexecs int) {
	defer utils.PrintTimeElapsed(time.Now(), "Total time spent:")
	ctx := context.Background()
	cl := dgraph.GetClient()

	for i := 0; i < outerloop; i++ {
		c := make(chan bool, parallelexecs)
		for j := 0; j < parallelexecs; j++ {
			go func(i, j int) {
				err := tryUpsert(ctx, i, j, cl)
				if err != nil {
					log.Fatal(err)
				}
				log.Println("Done from:", i, j)
				c <- true
			}(i, j)
		}

		for j := 0; j < parallelexecs; j++ {
			<-c
		}
		time.Sleep(100 * time.Millisecond)
		log.Println("All done")
	}
}
```

```
func tryUpsert(ctx context.Context, i, j int, cl *client.Dgraph) error {
	t := time.Now()
	defer utils.PrintTimeElapsed(t, "Elapsed time in tryUpsert:")
	txn := cl.NewTxn()
	defer txn.Discard(ctx)

	searchQ := fmt.Sprintf(`
					{
						find(func:eq(name, "Luke Skywalker_%v")){
							uid
						}
					}
				`, t)

	log.Println("Trying Search")
	_, err := txn.Query(ctx, searchQ)
	if err != nil {
		return err
	}

	q := fmt.Sprintf(`
						_:luke <name> "Luke Skywalker_%v" .
						_:sw1 <name> "Star Wars: Episode IV - A New Hope_%v" .
						_:sw1 <starring> _:luke .
					`, t, t)

	mu := &api.Mutation{SetNquads: []byte(q), IgnoreIndexConflict: true}
	_, err = txn.Mutate(ctx, mu)
	if err != nil {
		return err
	}
	log.Println("Trying commit")
	err = txn.Commit(ctx)

	return err
}
```

Use program with following values for **20 upserts in parallel** (apprx)
`outerloop=200000
parallelexecs=20`

- Expected behaviour and actual result.
Upsert call is taking 0.5 sec to 60 sec.
Load on the server is very high. Please find the output from the "top" attached:

```
top - 16:05:28 up 1 day,  3:06,  3 users,  load average: 74.17, 45.63, 29.42
Tasks: 160 total,   1 running, 159 sleeping,   0 stopped,   0 zombie
%Cpu(s): 96.6 us,  0.1 sy,  0.0 ni,  3.2 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:  15400948 total, 15236080 used,   164868 free,   136608 buffers
KiB Swap:        0 total,        0 used,        0 free.  1860984 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 4005 root      20   0 23.110g 0.012t  29176 S 773.1 81.7   6740:00 dgraph
 3985 root      20   0 3426252 354636  25660 S   0.7  2.3  22:16.95 dgraph
15794 ak170717  20   0   23696   1708   1164 R   0.3  0.0   0:00.01 top
    1 root      20   0   33780   2360    708 S   0.0  0.0   0:01.55 init
    2 root      20   0       0      0      0 S   0.0  0.0   0:00.00 kthreadd
    3 root      20   0       0      0      0 S   0.0  0.0   0:00.20 ksoftirqd/0
    5 root       0 -20       0      0      0 S   0.0  0.0   0:00.00 kworker/0:0H
    7 root      20   0       0      0      0 S   0.0  0.0   0:05.04 rcu_sched
    8 root      20   0       0      0      0 S   0.0  0.0   0:01.96 rcuos/0
    9 root      20   0       0      0      0 S   0.0  0.0   0:02.06 rcuos/1
   10 root      20   0       0      0      0 S   0.0  0.0   0:01.95 rcuos/2
   11 root      20   0       0      0      0 S   0.0  0.0   0:01.76 rcuos/3
   12 root      20   0       0      0      0 S   0.0  0.0   0:00.55 rcuos/4
   13 root      20   0       0      0      0 S   0.0  0.0   0:00.54 rcuos/5
   14 root      20   0       0      0      0 S   0.0  0.0   0:00.56 rcuos/6
   15 root      20   0       0      0      0 S   0.0  0.0   0:00.52 rcuos/7
   16 root      20   0       0      0      0 S   0.0  0.0   0:00.00 rcuos/8
   17 root      20   0       0      0      0 S   0.0  0.0   0:00.00 rcuos/9
   18 root      20   0       0      0      0 S   0.0  0.0   0:00.00 rcuos/10
   19 root      20   0       0      0      0 S   0.0  0.0   0:00.00 rcuos/11
   20 root      20   0       0      0      0 S   0.0  0.0   0:00.00 rcuos/12
   21 root      20   0       0      0      0 S   0.0  0.0   0:00.00 rcuos/13
```

Also the output of "vmstat 1":
```
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 4  0      0 205328 136404 1867164    0    0    19    60   26    0 51  1 48  0  0
128  0      0 205020 136404 1867164    0    0     0     0 7023 11421 83  0 16  0  0
 7  0      0 205020 136404 1867164    0    0     0     0 8326 14961 85  1 14  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 20624 37161 36  1 63  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 19766 35526 33  1 66  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 19708 35666 34  1 65  0  0
 4  0      0 205020 136404 1867164    0    0     0    48 20158 36026 32  2 66  0  0
 3  0      0 205020 136404 1867164    0    0     0     4 20052 35907 32  1 66  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 20017 36001 33  1 66  0  0
 5  0      0 205020 136404 1867164    0    0     0     0 19958 35992 32  1 67  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 20083 36089 32  1 67  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 19791 35445 31  1 68  0  0
 3  0      0 205020 136404 1867164    0    0     0    28 19896 35709 31  1 68  0  0
 2  0      0 205020 136404 1867164    0    0     0     0 20073 35867 31  1 68  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 20119 36251 31  1 68  0  0
 4  0      0 205020 136404 1867164    0    0     0     0 20029 36009 32  1 67  0  0
 6  0      0 205020 136404 1867164    0    0     0     0 19907 35782 30  1 68  0  0
 4  0      0 205020 136404 1867164    0    0     0     0 19673 35101 30  1 69  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 19624 34935 29  1 70  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 19343 34549 29  1 70  0  0
 2  0      0 205020 136404 1867164    0    0     0     0 19493 34602 30  1 69  0  0
 2  0      0 205020 136404 1867164    0    0     0     0 19240 34341 30  1 68  0  0
129  0      0 205020 136404 1867164    0    0     0     0 7871 13162 81  0 19  0  0
 4  0      0 205020 136404 1867164    0    0     0     0 7382 12480 87  1 12  0  0
 3  0      0 205020 136404 1867164    0    0     0     0 20513 37000 35  2 63  0  0
114  0      0 203916 136404 1867164    0    0     0  1556 15565 24810 55  1 44  1  0
63  0      0 205020 136404 1867276    0    0     0     0 2363 2862 100  0  0  0  0
52  0      0 205020 136404 1867276    0    0     0     0 2429 2939 99  0  1  0  0
23  0      0 205020 136404 1867276    0    0     0     0 2471 2908 99  0  1  0  0
129  0      0 205020 136404 1867276    0    0     0     0 2553 3210 100  0  0  0  0
35  0      0 204276 136404 1867276    0    0     0     0 2479 3135 100  0  0  0  0
 2  0      0 204028 136404 1867276    0    0     0     0 17774 31773 57  2 41  0  0
 2  0      0 204028 136404 1867276    0    0     0     8 19919 36203 36  1 63  0  0
 3  0      0 204028 136404 1867276    0    0     0     0 19879 36025 35  1 63  0  0
 4  0      0 204028 136404 1867276    0    0     0     0 19827 35730 34  2 65  0  0
 3  0      0 204028 136404 1867276    0    0     0     0 19588 35432 33  2 66  0  0
 2  0      0 204028 136404 1867276    0    0     0     4 19611 35415 32  1 67  0  0
 3  0      0 204028 136404 1867276    0    0     0    96 19689 35533 33  1 66  0  0
 2  0      0 204028 136404 1867276    0    0     0     0 19697 35296 33  2 66  0  0
 2  0      0 204028 136404 1867276    0    0     0     0 19496 35164 32  1 67  0  0
 3  0      0 204028 136404 1867276    0    0     0     0 19384 34838 31  1 68  0  0
 2  0      0 204028 136404 1867276    0    0     0     0 19625 35111 32  1 66  0  0
```

**Comment by ajayktkp** [comment]
The above stats are when dgraph is already having 1.5 mil nodes created with above program. I didn't check the stats earlier.

**Comment by peterstace** [comment]
> Upsert call is taking 0.5 sec to 60 sec

These numbers aren't good and are unexpected. We'll try to reproduce this result, and investigate what is causing the poor performance.

**Comment by janardhan1993** [comment]
@ajayktkp: One of the reasons for high cpu utilization is that you are trying to user term tokenizer for eq.
All of the mutations have a common term called luke. So all queries would need to do a second pass over the index uids to determine whether there is exact match or not. These are not cached also, so the server would end up doing millions of disk seeks for each query.

You should replace your schema with @index(exact) instead of @index(term).

**Comment by pawanrawal** [comment]
Duplicate of #1910

**Comment by ajayktkp** [comment]
Thanks for the quick response. So, I stopped my program and restarted the dgraph too. The load is still the same. Then I ran the mutation query to change the schema which is running for past 1 hr and still not returned and load is still high (top - 14:02:40 up 2 days,  1:04,  2 users,  load average: 122.71, 122.27, 121.01). Total nodes in graph are around 2Mil. What do you suggest to do here?

**Comment by ajayktkp** [comment]
Please don't close the issue yet as I'm not able to still stabilize my DGraph server.

**Comment by janardhan1993** [comment]
We are tracking this at 1910, so closed this one(As mentioned in earlier comment). Can you please retry from scratch.

**Comment by ajayktkp** [comment]
From scratch means, deleting everything and start with a fresh install?

**Comment by janardhan1993** [comment]
Yeah meanwhile we will fix the high load, changing the index should reduce the load but not sure whether it would fix the issue or not. Can you please try on a separate instance and update once.

**Comment by ajayktkp** [comment]
cool. thanks 👍
